src/main/java/io/numaproj/numaflow/reducestreamer/ReduceSupervisorActor.java

```diff
@@ -0,0 +1,196 @@
+        shutdownActor.tell(reason, ActorRef.noSender());
+        Service.reduceActorSystem.stop(getSelf());
+    }
+
+    @Override
+    public SupervisorStrategy supervisorStrategy() {
+        return new ReduceSupervisorStrategy();
+    }
+
+
+    @Override
+    public void postStop() {
+        log.debug("post stop of supervisor executed - {}", getSelf().toString());
+        shutdownActor.tell(Constants.SUCCESS, ActorRef.noSender());
+    }
+
+    @Override
+    public Receive createReceive() {
+        return ReceiveBuilder
+                .create()
+                .match(ActorRequest.class, this::invokeActors)
+                .match(String.class, this::sendEOF)
+                .match(ActorResponse.class, this::handleActorResponse)
+                .build();
+    }
+
+    /*
+        based on the keys of the input message invoke the right reduce streamer actor
+        if there is no actor for an incoming set of keys, create a new reduce streamer actor
+        track all the child actors using actors map
+     */
+    private void invokeActors(ActorRequest actorRequest) {
+        String[] keys = actorRequest.getKeySet();
+        String uniqueId = actorRequest.getUniqueIdentifier();
+        if (!actorsMap.containsKey(uniqueId)) {
+            ReduceStreamer reduceStreamerHandler = reduceStreamerFactory.createReduceStreamer();
+            ActorRef actorRef = getContext()
+                    .actorOf(ReduceStreamerActor.props(
+                            keys,
+                            this.md,
+                            reduceStreamerHandler,
+                            this.responseStreamActor));
+            actorsMap.put(uniqueId, actorRef);
+        }
+
+        HandlerDatum handlerDatum = constructHandlerDatum(actorRequest.getRequest().getPayload());
+        actorsMap.get(uniqueId).tell(handlerDatum, getSelf());
+    }
+
+    private void sendEOF(String EOF) {
+        for (Map.Entry<String, ActorRef> entry : actorsMap.entrySet()) {
+            entry.getValue().tell(EOF, getSelf());
+        }
+    }
+
+    private void handleActorResponse(ActorResponse actorResponse) {
+        if (actorResponse.getType() == ActorResponseType.EOF_RESPONSE) {
+            // forward the response to the response stream actor to send back to gRPC output stream.
+            this.responseStreamActor.tell(actorResponse, getSelf());
+        } else if (actorResponse.getType() == ActorResponseType.READY_FOR_CLEAN_UP_SIGNAL) {
+            // the corresponding actor is ready to be cleaned up.
+            // remove the child entry from the map.
+            // if there are no entries in the map, that means processing is
+            // done we can close the entire stream.
+            actorsMap.remove(actorResponse.getActorUniqueIdentifier());
+            if (actorsMap.isEmpty()) {
+                responseObserver.onCompleted();
+                getContext().getSystem().stop(getSelf());
+            }
+        } else {
```

Comment:
As per our offline discussion, let's centralise response observer tasks in one actor.  We can send EOF to the response actor when the map is empty, and trigger `onCompleted()` inside the response actor.